sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/implementation/FluxInputStream.java

```diff
@@ -0,0 +1,242 @@
+import com.azure.core.util.FluxUtil;
+import com.azure.core.util.logging.ClientLogger;
+import org.reactivestreams.Subscription;
+import reactor.core.publisher.Flux;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * An InputStream interface that subscribes to a Flux allows data to be read.
+ */
+public class FluxInputStream extends InputStream {
+
+    private ClientLogger logger = new ClientLogger(FluxInputStream.class);
+
+    // The data to subscribe to.
+    private Flux<ByteBuffer> data;
+
+    // Subscription to request more data from as needed
+    private Subscription subscription;
+
+    private ByteArrayInputStream buffer;
+
+    private boolean subscribed;
+    private boolean fluxComplete;
+    private boolean waitingForData;
+
+    /* The following lock and condition variable is to synchronize access between the reader and the
+        reactor thread asynchronously reading data from the Flux. If no data is available, the reader
+        acquires the lock and waits on the dataAvailable condition variable. Once data is available
+        (or an error or completion event occurs) the reactor thread acquires the lock and signals that
+        data is available. */
+    private final Lock lock;
+    private final Condition dataAvailable;
+
+    private IOException lastError;
+
+    /**
+     * Creates a new FluxInputStream
+     *
+     * @param data The data to subscribe to and read from.
+     */
+    public FluxInputStream(Flux<ByteBuffer> data) {
+        this.subscribed = false;
+        this.fluxComplete = false;
+        this.waitingForData = false;
+        this.data = data;
+        this.lock = new ReentrantLock();
+        this.dataAvailable = lock.newCondition();
+    }
+
+    @Override
+    public int read() throws IOException {
+        byte[] ret = new byte[1];
+        int count = read(ret, 0, 1);
+        return count == -1 ? -1 : ret[0];
+    }
+
+    @Override
+    public int read(byte[] b, int off, int len) throws IOException {
+        validateParameters(b, off, len);
+
+        /* If len is 0, then no bytes are read and 0 is returned. */
+        if (len == 0) {
+            return 0;
+        }
+        /* Attempt to read at least one byte. If no byte is available because the stream is at end of file,
+           the value -1 is returned; otherwise, at least one byte is read and stored into b. */
+
+        /* Not subscribed? subscribe and block for data */
+        if (!subscribed) {
+            blockForData();
+        }
+        /* Now, we have subscribed. */
+        /* At this point, buffer should not be null. If it is, that indicates either an error or completion event
+           was emitted by the Flux. */
+        if (this.buffer == null) { // Only executed on first subscription.
```

Comment:
I think you may want to lock around checking if buffer is null and assigning to it here. You lock around that in your onNext below, so modifying it here without locking will give you a race condition, i think

Reply:
Same thing wherever you access this.buffer

---
In reply to: [425432427](https://github.com/Azure/azure-sdk-for-java/pull/10641#discussion_r425432427) [](ancestors = 425432427)

Reply:
Do you think it'd be ok if I just move the buffer population above the locking logic? The intent was for the waitingForData variable to be protected by the lock moreso than the buffer

Reply:
Uhh. It feels safest to put the buffer population in the lock because what if you get two items emitted really quickly and they have different amounts of data or one of them is the end of the stream and is actually empty. Then we may have entered an if block based on one buffer but are operating on a different buffer for which the conditions are no longer true. But we can also walk through this offline and think about possibilities if you think it's safe to put the buffer stuff outside the lock

Reply:
We only request when we blockForData though so I'm not sure that could happen. We can talk about this next week